Prepare v0.59.0 release (#6447)
open-policy-agent/opa: CHANGELOG.md

All notable changes to this project will be documented in this file. This
project adheres to [Semantic Versioning](http://semver.org/).

## 0.59.0

This release adds tooling to help prepare existing policies for the upcoming OPA 1.0 release.

> This release also contains a mix of improvements, bugfixes and security fixes for third-party libraries.

Reviewer: ```suggestion
It also contains a mix of improvements, bugfixes and security fixes for third-party libraries.
```

> ### Rego v1

Author: I'm unsure about this section. Is this the time to inform the community about this, or should we wait until `opa fmt` also attempts to fix deprecated built-in calls?

We need get this out there sometime, though.

Reviewer: I think as long as we're clear that "the future release" is not this one, and we add some words such that this is unmistakably clear, it's good to get this out. :+1:

The future release of OPA 1.0 will introduce breaking changes to the Rego language. Most notably:

* the keywords that must be imported through `import future.keywords` into a module before use will be part of the Rego language by default, without the need to import them.
* the `if` keyword will be required before the body of a rule.
* the `contains` keyword will be required when declaring a multi-value rule (partial set rule).
* deprecated built-in functions will be removed.

This release introduces a new `--rego-v1` flag to the `opa fmt` and `opa check` commands.

When used with `opa fmt`, the `--rego-v1` flag will format the module(s) according to the new Rego syntax of OPA 1.0. 
Formatted modules are compatible with both the current version of OPA and 1.0. 
Modules using deprecated built-ins will terminate formatting with an error. Future versions of OPA will support rewriting applicable calls with equivalent Rego compatible with 1.0. 

> When used with `opa check`, the `--rego-v1` flag will check that the modules are compatibility with both the current version of OPA and 1.0.

Reviewer 2: ```suggestion
When used with `opa check`, the `--rego-v1` flag will check that the modules are compatible with both the current version of OPA and 1.0.
```
[typo]

#### Relevant Changes

- cmd: Adding `--rego-v1` flag to `check` cmd ([#6429](https://github.com/open-policy-agent/opa/issues/6429)) authored by @johanfylling
- cmd & format: Adding rego-v1 mode to `opa fmt` ([#6297](https://github.com/open-policy-agent/opa/issues/6297)) authored by @johanfylling
- ast: Adding capability feature for the `rego.v1` import (#6375) (authored by @johanfylling)
- ast: Skip if keyword requirement for default rule (`rego.v1`) ([#6356](https://github.com/open-policy-agent/opa/pull/6356)) authored by @ashutosh-narkar
- rego.v1: Fixing erroneous missing value assignment error ([#6364](https://github.com/open-policy-agent/opa/issues/6364)) authored by @johanfylling
- rego.v1: Improving support for rules with chained bodies ([#6370](https://github.com/open-policy-agent/opa/issues/6370)) authored by @johanfylling
- ast: Add `rego.v1` import ([#6247](https://github.com/open-policy-agent/opa/issues/6247)) introduced in OPA 0.58.0, authored by @johanfylling

### Runtime, Tooling, SDK

- ast: Adding `rule_head_refs` capabilities feature flag ([#6334](https://github.com/open-policy-agent/opa/issues/6334)) authored by @johanfylling

> - build: Remove rootless image variant ([#4295](https://github.com/open-policy-agent/opa/issues/4295)) authored by @ashutosh-narkar

Reviewer: This is important to call out once more, isn't it? 🤔

  - docs: Fixed XACML Policy in documentation (Comparing to Other Systems) to be XACML 3.0 compliant ([#6438](https://github.com/open-policy-agent/opa/pull/6438)) authored by @cdanger
  - docs: Update docs on rego.v1 / OPA 1.0 ([#6365](https://github.com/open-policy-agent/opa/pull/6365)) authored by @anderseknert
  - docs: Update spinnaker integration ([#6414](https://github.com/open-policy-agent/opa/pull/6414)) authored by @charlieegan3
  - docs: Add legitify to ecosystem ([#6369](https://github.com/open-policy-agent/opa/pull/6369)) authored by @charlieegan3
  - docs: add cheat sheet link ([#6362](https://github.com/open-policy-agent/opa/pull/6362)) authored by @charlieegan3
  - docs: add newstack blog to regal ([#6372](https://github.com/open-policy-agent/opa/pull/6372)) authored by @charlieegan3
  - docs: Disk storage broken link ([#6425](https://github.com/open-policy-agent/opa/pull/6425)) authored by @francoisauclair911
  - docs: Update istio envoy tutorial to use AuthorizationPolicy ([#6426](https://github.com/open-policy-agent/opa/pull/6426)) authored by @tjons
- Dependency updates; notably:

>   - golang: Update golang to 1.21.4 (authored by @ashutosh-narkar)

Reviewer: I think only this one and the OTel version is notable, here. What do you think? Would you care much about the rest of the bumps?

  - github.com/containerd/containerd from 1.7.7 to 1.7.9
  - github.com/go-logr/logr from 1.2.4 to 1.3.0
  - github.com/google/uuid from 1.3.1 to 1.4.0
  - github.com/gorilla/mux from 1.8.0 to 1.8.1
  - github.com/spf13/cobra from 1.7.0 to 1.8.0
  - golang.org/x/net from 0.17.0 to 0.19.0
  - golang.org/x/time from 0.3.0 to 0.5.0
  - go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp from 0.45.0 to 0.46.1
  - go.opentelemetry.io/otel/exporters/otlp/otlptrace from 1.19.0 to 1.21.0

>   - go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc from 1.19.0 to 1.21.0

Reviewer: It's combine these, sayng "OpenTelemetry (contrib) 1.21.0/0.46.1"
